### PR TITLE
Ifpack2,Xpetra,MueLu: Prep for high-order Maxwell

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/muelu @trilinos/ifpack2 @trilinos/xpetra 

## Motivation
Ifpack2: 
- Cache temp vectors in Hiptmair smoother.

Xpetra-Thyra:
- Add conversion of diagonal Thyra operators.
- Convert to `Xpetra::Matrix` instead of `CrsMatrix`. (@searhein That impacted FROSch.)

MueLu Maxwell1:
- Add a Stratimikos adapter.
- Allow passing in user-computed matrices.
- Add RefMaxwell coarse solve option.
- Fix a couple of smaller issues.

### CODE REVIEW

File: packages/muelu/adapters/xpetra/MueLu_OperatorSmoother_def.hpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+
+
+namespace MueLu {
+
+  template <class Scalar,class LocalOrdinal, class GlobalOrdinal, class Node>
+  OperatorSmoother<Scalar, LocalOrdinal, GlobalOrdinal, Node>::OperatorSmoother(const std::string type, const Teuchos::ParameterList& paramList)
+  : type_(type)
+  {
+    const bool solverSupported = (type_ == "RefMaxwell");
```
[the remark this answers is not in the document]
Comment (cgcgcg) [reply packages/muelu/adapters/xpetra/MueLu_OperatorSmoother_def.hpp:71]: It is true that for now, this only supports RefMaxwell. The idea was to be able to use any `Xpetra::Operator` as a smoother/coarse solver. Not sure we will ever need that though. But I can definitely rename this to `RefMaxwellSmoother`. Would that be less confusing?

File: packages/muelu/adapters/xpetra/MueLu_OperatorSmoother_decl.hpp [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+    //! @name Computational methods.
+    //@{
+
+    /*! @brief Set up the smoother.
+
+    This creates the underlying Operator smoother object, copies any parameter list options
+    supplied to the constructor to the Operator object, and computes the preconditioner.
+    */
+    void Setup(Level &currentLevel);
```
[the remark this answers is not in the document]
Comment (cgcgcg) [reply packages/muelu/adapters/xpetra/MueLu_OperatorSmoother_decl.hpp:116]: I believe MueLu normally uses `Setup`?

File: packages/muelu/adapters/xpetra/MueLu_OperatorSmoother_def.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jhux2) [reply packages/muelu/adapters/xpetra/MueLu_OperatorSmoother_def.hpp:71]: My understanding is that an OperatorSmoother is a solver that depends only on the action of the matrix.  In the current use case, the solver is RefMaxwell, but as you say, other methods that don't need access to matrix entries would be also be valid.

[the remark this answers is not in the document]
Comment (jhux2) [reply packages/muelu/adapters/xpetra/MueLu_OperatorSmoother_def.hpp:71]: So it was really the name `OperatorSmoother` that threw me.  But if my understanding above is correct, then I'm fine with this, and there's no need to change anything.

[the remark this answers is not in the document]
Comment (cgcgcg) [reply packages/muelu/adapters/xpetra/MueLu_OperatorSmoother_def.hpp:71]: Yes, that understanding is correct :-)

**Review by jhux2 — APPROVED** [review]
lgtm

**Pull request merged**